### DISCUSSION TITLE
Supporting Schema-defined Validation Arguments

### DISCUSSION THREAD

**Discussion by Coffee2Bits** [DevOps-MBSE/AaC#669] [Ideas]
This discussion captures research and ideas on how to allow users to define complex validator input and how we might validate that validation input arguments are consistent with the defined input schema.

## Background
Before getting into the meat and potatoes of the proposed changes, I want to touch on the current methods and constraints to elaborate on why some of the proposed changes are necessary and how we can navigate some short-comings of the AaC DSL.

### Current Method for Defining Validator Input
Currently we define validator arguments as a [list of strings in the core spec](https://github.com/DevOps-MBSE/AaC/blob/main/python/src/aac/spec/spec.yaml#L208-L210) which obviously limits users to only being able to leverage arguments that are a series of strings. There was an attempt to give users greater flexibility in the argument structure they can provide to validators in the [Pull Request Validator KW Args](https://github.com/DevOps-MBSE/AaC/pull/561). This PR sought to allow users to define a set of key-value pairs for greater flexibility in providing arguments to the validators, however, it can be limiting and doesn't allow user-defined validation of the arguments.

### Language Limitations
The AaC DSL and current way that we've defined validations for `schema` definitions let the use provide a list of `ValidationReference` as seen [here](https://github.com/DevOps-MBSE/AaC/blob/main/python/src/aac/spec/spec.yaml#L328-L331). These `ValidationReference` have the arguments defined [as a list of strings](https://github.com/DevOps-MBSE/AaC/blob/main/python/src/aac/spec/spec.yaml#L208-L210). The problem is that we can't dynamically define the argument structures per validation. The proposed solution is to define these arguments in the core spec as key-value pairs, unstructured data that the user can set arbitrarily. 

### Update the Validator Interface to Accept a List of Dictionaries
Leveraging the work done in [Pull Request Validator KW Args](https://github.com/DevOps-MBSE/AaC/pull/561), we can update the validator interface to accept a list of Python dictionaries as the set of arguments. Given that all YAML structures are ultimately parsed as dictionaries in Python, this will give our interface the flexibility to accept any user-defined data as arguments into the validator's implementation.

### Define Validator Argument Structures using Validation Behavior
Because of the way `ValidationReference` and the `validation` field in `schema` are defined, we can't allow users to easily define the expected data structure for validation arguments. The only suitable location for users to define the expected structure for validation arguments is in the `behavior` field of `Validation` definition. Currently, we hand-wave over the defined structure for validation arguments by re-using the generic structure `ValidatorInput` as seen in the `Type References Exist` validation [here](https://github.com/DevOps-MBSE/AaC/blob/main/python/src/aac/plugins/validators/defined_references/defined_references.yaml#L14-L16). Instead, users should be defining the actual structure of the expected arguments in the behavior input of the validator. However, we don't currently have a method of enforcing the defined structure and applying its own validations against the arguments provided.

Our example `Type references exist` validation currently looks like this:
```
validation:
  name: Type references exist
  description: Verify that references to definitions and models are defined.
  behavior:
    - name: Verify that referenced definitions are defined.
      type: REQUEST_RESPONSE
      description:
      input:
        - name: input
          type: ValidatorInput
      ...
```

Would instead define its arguments as a list of strings like so:
```
validation:
  name: Type references exist
  description: Verify that references to definitions and models are defined.
  behavior:
    - name: Verify that referenced definitions are defined.
      type: REQUEST_RESPONSE
      description:
      input:
        - name: input
          type: string[]
      ....
```

If we pretend that `Type references exist` has a more complex data structure for its arguments, we could define a schema and set that as our input type like so:
```
schema:
  name: MyValidationArgs
  fields:
    - name: Arg1
      type: int
    - name: Arg2
      type: string
---
validation:
  name: Type references exist
  description: Verify that references to definitions and models are defined.
  behavior:
    - name: Verify that referenced definitions are defined.
      type: REQUEST_RESPONSE
      description:
      input:
        - name: input
          type: MyValidationArgs
      ....
```

### Enforce Argument Data Structures Via New Validators
In order to support some structural and content validation of the arguments passed to a validation, we can develop a new validation `Validation arguments match validation behavior input`. This validator will cover the missing gap in the validation system and would look up the type defined by the `Validation`'s `behavior:input` field. 
#### How do We Identify the Correct Validator Behavior and Input?
One of the potential shortfalls of the newly proposed validator is that there is nothing stopping a user from defining multiple behaviors for a validator leading to the question of how would we determine which behavior and input to select when verifying the structure of the arguments? There are a couple approaches I can think of: 1. limit validations to one behavior and one input, 2. enforce that validations have a behavior named `validate` with an input named `input` or `arguments`. 

I'm pretty 50/50 on which approach I'd want to take, but I find myself slightly partial to the the first approach of limiting the behaviors and inputs defined for a `Valdiation` definition. This allows users to maintain flexibility while preventing us from having to programmatically infer the user's intent and make a choice for them. With this slight preference in mind, I will only propose a story for this approach; if we decide that the second approach is preferable then I can write a proposed story for that approach instead.

#### Update the Validators to Accept Keyword Arguments
This story will largely be to bring the changes in the [Pull Request Validator KW Args](https://github.com/DevOps-MBSE/AaC/pull/561) into the main branch with the additional removal of `*validation_args` from the [validator interface](https://github.com/DevOps-MBSE/AaC/blob/main/python/src/aac/plugins/validators/defined_references/_validate_references.py#L19). We may also want to adjust the PR changes to have the arguments be a list of key-value pairs. This will be funky, but if we equate a key-value pair to an entry of any structure then we will need a list of key-value pairs to allow users to be able to use a list of schema-defined arguments.

Sample AC:
* All existing validator interfaces are updated so that the argument `validation_args` is a list[dict].
* The [validation](https://github.com/DevOps-MBSE/AaC/blob/ca4dd5713d97a22fe150a3525df668c7a8cd61fb/python/src/aac/validate/_validate.py#L175-L177) is updated to extract arguments as a `list[dict]` instead of a list of strings
* There are tests demonstrating the validation working for:
	* A Validation with no arguments
	* A Validation with a list of strings as arguments
	* A Validation with a list of key-value pairs (dict)

#### Create the 'Validation arguments match validation behavior input' Validator
This story will focus on creating a validator that performs a look up on the `Validation` definition's `behavior:input` field `type` and use the referenced `schema` definition there as the `schema` definition by which to validate the argument input (or treat is a primitive validation if the field is defined as a primitive). This will require some modifications to the internal functions in [`aac.validate._validate`](https://github.com/DevOps-MBSE/AaC/blob/main/python/src/aac/validate/_validate.py); likely replicating the function [`_validate_definition`](https://github.com/DevOps-MBSE/AaC/blob/ca4dd5713d97a22fe150a3525df668c7a8cd61fb/python/src/aac/validate/_validate.py#L130-L156), however, instead of passing in a definition as the first argument we will pass in a list of dictionaries (our validation arguments) and we would need to add another argument for the `schema` definition that defines the structure of the validation arguments. This validator would effectively take the schema/type defined in the validation's behavior input, massage the validation argument data, then run the data and its 'new' schema back through our validation system.

I can't think of a more native way to support this, and this validator will be implementing some anti-patterns such as accessing and calling internal validation package functions. I'd be happy to discuss further and see if we can come up with a better way to support the dynamic validation of validator arguments.

Sample AC:
* There is a new Validator created named 'Validation arguments match validation behavior input'
* The new validator is applied to the `Schema` definition
* The validator does not accept any arguments
* The validator returns an error if the validation arguments do not match the type specified in the validation's `behavior:input`
* The validator returns an error if the validation has arguments but there is no input defined in the validation's `behavior:input`
* The validator returns an error if the validation has no arguments but there is input defined in the validation's `behavior:input`
* The validator applies validation to each of the validation arguments

#### Create the 'Validation only has one behavior and up to one input' Validator
This story will have a validator created that forces `Validation` definitions to have one and only one behavior entry accompanied with 0 or 1 input entries. This limit will allow us to go with the first `Validation` `behavior` entry and the first `input` of the `behavior` if the behavior has inputs. 

Sample AC:
* There is a new Validator created named 'Validation only has one behavior and up to one input'
* The new validator is applied to the `Validation` definition
* The validator does not accept any arguments
* The validator returns an error if a `Validation` definition has no behavior entries
* The validator returns an error if a `Validation` definition has more than one behavior entry
* The validator returns an error if the `Validation`'s `behavior` entry has more than one input

**Comment by jondavid-black** [comment]
Looking at the aac core spec, I'm a bit confused about this discussion.  The validator schema simply asks that a used define a behavior which in turn provides the ability to provide input and output definitions.  These input and output are both just Field types, allowing the user to define a list of named and typed entries.  The spec definition seems to support the desired behavior as-is.

But I can also see that existing validators seem to use the ValidatorInput and ValidatorOutput types for input and output respectively.  This appears to be a matter of convention per the core spec, but does seem to be applied universally.

Am I interpreting this correctly to be a change to eliminate the convention of ValidatorInput and ValidatorOutput and simply implement the validator schema currently in core?  If so, I completely support that change!  In fact, I would support eliminating the ValidatorInput and ValidatorOutput schemas all together to avoid potential confusion.

I can understand the desire to constrain a validator to a single behavior.  I would simplify the recommended story to just change the core spec to so that the behavior field in validator changes the type from ```Behavior[]``` to ```Behavior```.  I would recommend we make no other changes, allowing inputs to be defined however the plug-in developer deems appropriate for their need.

**Reply by Coffee2Bits** [reply]
Unfortunately the spec doesn't allow us to dynamically define the argument input structure.

The core spec defines the validation arguments as:
```yaml
schema:
  name: ValidationReference
  fields:
    - name: name
      type: reference
      description: The name of the validation definition.
    - name: arguments
      type: string[]
      description: A list of arguments to pass to the validator.
```

The field `arguments` only allows us to define arguments as a list of strings, but there are instances where we may want to allow users to provide more complex argument inputs. The type defined by the behavior inputs of validations doesn't affect the structure of the arguments.

```
Am I interpreting this correctly to be a change to eliminate the convention of ValidatorInput and ValidatorOutput and simply implement the validator schema currently in core? If so, I completely support that change! In fact, I would support eliminating the ValidatorInput and ValidatorOutput schemas all together to avoid potential confusion.
```

Correct. We want `ValidatorInput` replaced with per-validation specification of the arguments. `ValidatorOutput` can probably still be generic since we do have a standard class for reporting the findings.

```
I can understand the desire to constrain a validator to a single behavior. I would simplify the recommended story to just change the core spec to so that the behavior field in validator changes the type from Behavior[] to Behavior.
```
That's a good point, however, we would still need to limit the inputs since behavior defines input as `Field[]`.

**Reply by Coffee2Bits** [reply]
Now that I'm thinking on this some more, the `ValidatorInput` corresponds to the validator interface. Depending on how closely we want to model the input to our validators, we will need to account for the definition, schema, and context alongside the args that comprise the overall `ValidatorInput`. We could  break this out into the separate components for the behavior input such that `Validation` has a `Behavior` field with expected input entries for `definition`, `schema_definition`, `context`, and `arguments` where `arguments` defines the structure for the arguments.

**Comment by lizzcondrey** [comment]
I'm starting to get confused on the intent here now as well for modifying the `ValidatorInput`. It's pushed me towards a few questions and needs of clarification:

1. Is modifying the `ValidatorInput` still all this is trying to address? Is sounding like maybe it has expanded, or I am just confusing myself.
2.  What does this change for how a validator is constructed? (Have a model example?) Because it sounds like it would change more than just the original proposed modification for validator definition. 
3. What would the components (`definition`, `schema_definition`, `context`, and `arguments`) provide towards functionality in comparison to what is already there? Also, how do they relate to what is already part of a validator? It sounds like it is just making the `input` clearer on what is being provided within that `input`?

**Reply by Coffee2Bits** [reply]
>Is modifying the ValidatorInput still all this is trying to address? Is sounding like maybe it has expanded, or I am just confusing myself.

Yes, it would still modify it. The basic goal is to replace `ValidatorInput` with something user-defined.

> What does this change for how a validator is constructed? (Have a model example?) Because it sounds like it would change more than just the original proposed modification for validator definition.

Validators would still operate the same except for the final argument in the interface `*validation_args` [(example)](https://github.com/DevOps-MBSE/AaC/blob/9570780a808655a343bb31ecc343b814cc4ab2c6/python/src/aac/plugins/validators/reference_targets/_validate_reference_targets.py#L19) being altered to a key-value map to support generic yaml structures. We may also need to include another argument that's the structure of the validation args, but that's implementation minutia. 

> What would the components (definition, schema_definition, context, and arguments) provide towards functionality in comparison to what is already there? Also, how do they relate to what is already part of a validator? It sounds like it is just making the input clearer on what is being provided within that input?

My comment was an aside on the fidelity with which we want to model our validation. If we wanted to model it 100% for future use with generating behavior tests then we should consider modeling the validator interface completely. If we are only interested in a low-fidelity model then we can simply replace `ValidatorInput` with whatever the user defines the validation arguments as.

**Reply by lizzcondrey** [reply]
>  Yes, it would still modify it. The basic goal is to replace `ValidatorInput` with something user-defined.
> 
And by user-defined you mean to try and allow the more granularity of what makes up the `ValidatorInput`?

> > Update the Validators to Accept Keyword Arguments
> >
> > This story will largely be to bring the changes in the https://github.com/DevOps-MBSE/AaC/pull/561 into the main branch with the additional removal of *validation_args from the [validator interface](https://github.com/DevOps-MBSE/AaC/blob/main/python/src/aac/plugins/validators/defined_references/_validate_references.py#L19). We may also want to adjust the PR changes to have the arguments be a list of key-value pairs. This will be funky, but if we equate a key-value pair to an entry of any structure then we will need a list of key-value pairs to allow users to be able to use a list of schema-defined arguments.
> >
> Validators would still operate the same except for the final argument in the interface `*validation_args` [(example)](https://github.com/DevOps-MBSE/AaC/blob/9570780a808655a343bb31ecc343b814cc4ab2c6/python/src/aac/plugins/validators/reference_targets/_validate_reference_targets.py#L19) being altered to a key-value map to support generic yaml structures. We may also need to include another argument that's the structure of the validation args, but that's implementation minutia.
>
Guess I am still not really following what role the `*validation_args` serves, I don't think I have found a validator that makes use of it yet? Can you give me an example of one that does or more information to explain why both removal or alteration of the `*validation_args` is suggested above? 

> My comment was an aside on the fidelity with which we want to model our validation. If we wanted to model it 100% for future use with generating behavior tests then we should consider modeling the validator interface completely. If we are only interested in a low-fidelity model then we can simply replace `ValidatorInput` with whatever the user defines the validation arguments as.
>
So, saying that we could present all of the interface components to the user at the definition level rather than only within the code?
